File: Python/DP/M-416. Partition Equal Subset Sum.py
class Solution:
    @classmethod
    def canPartition(self, nums: list[int]) -> bool:
        n = len(nums)
        ttl_num = sum(nums)
        if ttl_num % 2 != 0:
            return False

        half = ttl_num // 2
        dp = [[-1 for j in range(half+1)] for i in range(n)]

        for rem_wt in range(half + 1):
            first_item_val = nums[0]
            dp[0][rem_wt] = first_item_val if rem_wt >= first_item_val else 0

        for item_cnt in range(1, n):
            item_val = nums[item_cnt]
            for rem_wt in range(half + 1):
                not_picked_sum = dp[item_cnt - 1][rem_wt]
                picked_sum = dp[item_cnt - 1][rem_wt - item_val] + \
                    item_val if rem_wt >= item_val else 0
                dp[item_cnt][rem_wt] = max(not_picked_sum, picked_sum)

        return dp[n-1][half] == half


test_data = [
    {
        "nums": [5, 1, 11, 5],
        "ans": True,
    },
    {
        "nums": [1, 2, 3, 5],
        "ans": False,
    },
    {
        "nums": [1, 2, 5],
        "ans": False,
    },
]

for data in test_data:
    print(data["ans"] == Solution.canPartition(data["nums"]))

File: DP/M-416. Partition Equal Subset Sum/M-416. Partition Equal Subset Sum.ts
// one dimension array
// function canPartition(nums: number[]): boolean {
//   const n = nums.length;
//   const sum = nums.reduce((sum, cur) => sum + cur, 0);

//   /* if sum is odd number */
//   if (sum % 2 !== 0) return false;

//   /*
//    * half is the target we need to find if there is any combination whose
//    * sum are equal to it.
//    **/
//   const half = sum / 2;
//   const dp = new Array(half + 1).fill(-1);

//   for (let i = 0; i < n; i++) {
//     const itemVal = nums[i];
//     for (let remWt = half; remWt <= 0; remWt--) {
//       const notPickedSum = dp[remWt];
//       const pickedSum = remWt >= itemVal ? dp[remWt - itemVal] + itemVal : 0;
//       dp[remWt] = Math.max(notPickedSum, pickedSum);
//     }
//   }
//   return dp[half] === half;
// }

// two dimensions array
function canPartition(nums: number[]): boolean {
  const n = nums.length;
  const sum = nums.reduce((sum, cur) => sum + cur, 0);

  /* if sum is odd number */
  if (sum % 2 !== 0) return false;

  /*
   * half is the target we need to find if there is any combination whose
   * sum are equal to it.
   **/
  const half = sum / 2;

  /*
   * store the maximum sum for picked items with each remaining weight
   * including zero (no picked item)
   **/
  const dp = new Array(n).fill(null).map((_) => new Array(half + 1));
  /* pick the 1st item to initialize the dp array */
  for (let remWt = 0; remWt < half + 1; remWt++) {
    const firstItemVal = nums[0];
    dp[0][remWt] = remWt >= firstItemVal ? firstItemVal : 0;
  }
  /* pick the rest items starting from 1 since the 1st item has already been picked */
  for (let itemCnt = 1; itemCnt < n; itemCnt++) {
    const itemVal = nums[itemCnt];
    for (let remWt = 0; remWt < half + 1; remWt++) {
      const notPickedSum = dp[itemCnt - 1][remWt];
      const pickedSum =
        remWt >= itemVal ? dp[itemCnt - 1][remWt - itemVal] + itemVal : 0;
      /* update dp */
      dp[itemCnt][remWt] = Math.max(notPickedSum, pickedSum);
    }
  }
  return dp[n - 1][half] === half;
}

const testData = [
  {
    nums: [5, 1, 11, 5],
    ans: true,
  },
  {
    nums: [1, 2, 3, 5],
    ans: false,
  },
  {
    nums: [1, 2, 5],
    ans: false,
  },
];

for (const data of testData) {
  console.log(data.ans === canPartition(data.nums));
}
